src/backend/distributed/executor/intermediate_results.c

```diff
@@ -795,8 +795,32 @@ ReadIntermediateResultsIntoFuncOutput(FunctionCallInfo fcinfo, char *copyFormat,
 		int statOK = stat(resultFileName, &fileStat);
 		if (statOK != 0)
 		{
-			ereport(ERROR, (errcode_for_file_access(),
-							errmsg("result \"%s\" does not exist", resultId)));
+			/*
+			 * When the file does not exist, it could mean two different things.
+			 * First -- and a lot more common -- case is that a failure happened
+			 * in a concurrent backend on the same distributed transaction. And,
+			 * one of the backends in that transaction has already been roll
+			 * backed, which has already removed the file. If we throw an error
+			 * here, the user might see this error instead of the actual error
+			 * message. Instead, we prefer to WARN the user and pretend that the
+			 * file has no data in it. In the end, the user would see the actual
+			 * error message for the failure.
+			 *
+			 * Second, in case of any bugs in intermediate result broadcasts,
+			 * we could try to read a non-existing file. That is most likely
+			 * to happen during development. Thus, when asserts enabled, we throw
+			 * an error instead of WARNING so that the developers cannot miss.
+			 */
+			int logLevel = WARNING;
+
+#ifdef USE_ASSERT_CHECKING
+			logLevel = ERROR;
+#endif
+			ereport(logLevel, (errcode_for_file_access(),
+							   errmsg("Query could not find the intermediate result file "
+									  "\"%s\", it was mostly likely deleted due to an "
+									  "error in a parallel process within the same "
+									  "distributed transction", resultId)));
 		}
 
 		ReadFileIntoTupleStore(resultFileName, copyFormat, tupleDescriptor, tupleStore);
```

Comment:
What does this function do for a non-existent file?

Reply:
Ah,  of course, with assert  enabled, we exit  the function so I missed  that.